feat: rewrite docs generation

lukekarrys opened the pull request:
TODO:
- [x] Test snapshots for all generated docs content
- [x] Gitignore only bundled deps

lukekarrys commented on package.json:
```diff
@@ -225,7 +225,8 @@
     "prelint": "rimraf test/npm_cache*",
-    "resetdeps": "bash scripts/resetdeps.sh"
+    "resetdeps": "bash scripts/resetdeps.sh",
+    "prepack": "npm run docs"
```
Docs are now generated via `prepack` instead of being part of the `Makefile`.

lukekarrys commented on .github/workflows/create-cli-deps-pr.yml:
```diff
@@ -64,7 +64,6 @@ jobs:
           cd cli
           git checkout "$npm_vtag"
-          make
           make release
```
Removing this was due to the old default target being `make all` which just ran `make docs`. Docs are now built via `prepack` so when `scripts/release.sh` calls `pack` the docs will get built.

lukekarrys commented:
Since the man and html docs were generated, and we are no longer checking in the generated markdown docs, I did a full diff between the docs in `8.17.0` and in this branch BEFORE I removed all the extra placeholders. The full diff is below and it shows that we are now correctly replacing the versions, adding a simple synopsis for the `npm` command, and adding two copied man files.

<details><summary>Diff from 8.17.0.tgz</summary>
<p>

```
diff --git a/package/man/.DS_Store b/man/.DS_Store
index 63b47e9ed..73cb601a0 100644
Binary files a/package/man/.DS_Store and b/man/.DS_Store differ
diff --git a/package/man/man1/npm.1 b/man/man1/npm.1
index 04133f580..5f2b4a706 100644
--- a/package/man/man1/npm.1
+++ b/man/man1/npm.1
@@ -2,6 +2,12 @@
 .SH "NAME"
 \fBnpm\fR \- javascript package manager
 .SS Synopsis
+.P
+.RS 2
+.nf
+npm
+.fi
+.RE
 .SS Version
 .P
 8\.17\.0

diff --git a/man/man5/npm-global.5 b/man/man5/npm-global.5
new file mode 100644
index 000000000..b6efd3c88
--- /dev/null
+++ b/man/man5/npm-global.5
@@ -0,0 +1,226 @@

diff --git a/man/man5/npm-json.5 b/man/man5/npm-json.5
new file mode 100644
index 000000000..d4f972311
--- /dev/null
+++ b/man/man5/npm-json.5
@@ -0,0 +1,1296 @@

diff --git a/docs/content/.DS_Store b/docs/content/.DS_Store
new file mode 100644
index 000000000..22e638cd1
Binary files /dev/null and b/docs/content/.DS_Store differ

diff --git a/package/docs/content/commands/npm-ls.md b/docs/content/commands/npm-ls.md
index a7936fafc..64719f22c 100644
--- a/package/docs/content/commands/npm-ls.md
+++ b/docs/content/commands/npm-ls.md
@@ -36,7 +36,7 @@ packages will *also* show the paths to the specified packages.  For
 example, running \`npm ls promzard\` in npm's source tree will show:
 
 \`\`\`bash
-npm@@VERSION@ /path/to/npm
+npm@8.17.0 /path/to/npm
 └─┬ init-package-json@0.0.4
   └── promzard@0.1.5
 \`\`\`

diff --git a/package/docs/content/commands/npm.md b/docs/content/commands/npm.md
index 18a68d03c..013459dce 100644
--- a/package/docs/content/commands/npm.md
+++ b/docs/content/commands/npm.md
@@ -7,11 +7,21 @@ description: javascript package manager
 ### Synopsis
 
 <!-- AUTOGENERATED USAGE DESCRIPTIONS START -->
+<!-- automatically generated, do not edit manually -->
+<!-- see lib/commands/npm.js -->
+
+\`\`\`bash
+npm
+\`\`\`
+
+<!-- automatically generated, do not edit manually -->
+<!-- see lib/commands/npm.js -->
+
 <!-- AUTOGENERATED USAGE DESCRIPTIONS END -->
 
 ### Version
 
-@VERSION@
+8.17.0
 
 ### Description
 
diff --git a/package/docs/content/using-npm/package-spec.md b/docs/content/using-npm/package-spec.md
index 0d3741018..fd64131fb 100644
--- a/package/docs/content/using-npm/package-spec.md
+++ b/docs/content/using-npm/package-spec.md
@@ -4,7 +4,6 @@ section: 7
 description: Package name specifier
 ---
 
-
 ### Description
 
 Commands like \`npm install\` and the dependency sections in the
diff --git a/docs/output/.DS_Store b/docs/output/.DS_Store
new file mode 100644
index 000000000..f3ab3515d
Binary files /dev/null and b/docs/output/.DS_Store differ
diff --git a/package/docs/output/commands/npm.html b/docs/output/commands/npm.html
index 523629237..d0bbf8682 100644
--- a/package/docs/output/commands/npm.html
+++ b/docs/output/commands/npm.html
@@ -148,6 +148,12 @@ npm command-line interface
 <div id="_content"><h3 id="synopsis">Synopsis</h3>
 <!-- raw HTML omitted -->
 <!-- raw HTML omitted -->
+<!-- raw HTML omitted -->
+<pre lang="bash"><code>npm
+</code></pre>
+<!-- raw HTML omitted -->
+<!-- raw HTML omitted -->
+<!-- raw HTML omitted -->
 <h3 id="version">Version</h3>
 <p>8.17.0</p>
 <h3 id="description">Description</h3>

```

<p></p>
</div></details>

lukekarrys commented on .github/workflows/ci.yml:
```diff
@@ -31,8 +31,8 @@ jobs:
           node-version: 16.x
           cache: npm
-      - run: make freshdocs
-      - run: node scripts/git-dirty.js
```
No docs are checked in so there isn't any need to check git status here anymore.

lukekarrys commented on package.json:
```diff
@@ -216,7 +216,7 @@
     "test-all": "npm run test --if-present --workspaces --include-workspace-root",
     "snap": "tap",
-    "postsnap": "make -s docs",
+    "docs": "npm run build -w docs",
```
Another spot where we don't have to run the docs command since they are all ignored and generated.

wraithgar commented on docs/lib/build.js:
```diff
@@ -0,0 +1,132 @@
+    readDocs(content),
+    readHtml(template),
+    readYaml(nav),
+    rmAll(man, html, md),
```
So this always removed all the docs every time, meaning that as long as we are running this the docs will be guaranteed to be built fresh? This solves an existing problem where `make` would sometimes not rebuild docs cause of timestamp shenanigans.

lukekarrys replied:
Yes it does everything from scratch each time. My only concern with this was how long it would take, so I profiled it with and without running each doc through `mdx`. It was ~3 seconds without doing mdx verification and ~5 seconds with it.

wraithgar commented:
How are we to review changes in usage output now? It was very handy having them in source control to see for PRs.  I suppose they still exist in the test snapshots for help/man?  Do the full man pages exist in snapshots?

lukekarrys commented:
> How are we to review changes in usage output now? It was very handy having them in source control to see for PRs. I suppose they still exist in the test snapshots for help/man? Do the full man pages exist in snapshots?

That's a good point. I just checked and we have snapshots for config's `describeAll()` and `shorthands`. I'm going to add snapshots that run for all commands to output their `usage` and `params`.

Also docs generation will fail if anything couldn't be replaced, so if something isn't being generated properly we will learn about it during CI (or if you run `npm run docs` locally).

lukekarrys committed:
feat: rewrite docs generation

High level overview of the changes here:

- The source for the docs content has moved from `docs/content/` to
  `docs/lib/content/`. The generated markdown is still written to
  `docs/content/` but that directory is now ignored from git.
- All generated content sections of the docs have been removed and
  replaced with single placeholder html comments such as `<!--
  AUTOGENERATED CONFIG DESCRIPTIONS -->`
- Placeholders are replaced with generated content only as part of the
  `prepack` step, so generated markdown is no longer checked in to
  source and all docs related `make` commands have been removed
- All docs (and docs related) snapshots have been moved to a single test
  file that outputs command usage and formats it with functions imported
  from `docs/lib/index.js`. So tests will fail if docs content changes
  until `npm run snap` is run.

lukekarrys commented:
~~There are a lot of commits here with bad messages, but this PR should be squashed.~~ So this is ready for review 😎 

Squashed this to add some deps commits that we can update to in the docs workspace, so this should now be rebased.

lukekarrys commented:
High level overview of the changes here:

- The source for the docs content has moved from `docs/content/` to `docs/lib/content/`. The generated markdown is still written to `docs/content/` but that directory is now ignored from git.
- All generated content sections of the docs have been removed and replaced with single placeholder html comments such as `<!-- AUTOGENERATED CONFIG DESCRIPTIONS -->`
- Placeholders are replaced with generated content only as part of the `prepack` step, so generated markdown is no longer checked in to source and all docs related `make` commands have been removed
- All docs (and docs related) snapshots have been moved to a single test file that outputs command usage and formats it with functions imported from `docs/lib/index.js`. So tests will fail if docs content changes until `npm run snap` is run.

lukekarrys committed:
chore: yaml@2.1.3, jsdom@20.0.1

Closes #5643
Closes #5642
Closes #5630
Closes #5638
